index.d.ts

```diff
@@ -0,0 +1,31 @@
+declare module 'objection-soft-delete' {
+  import { Model } from 'objection';
+
+  class SDQueryBuilder<M extends Model> {
+    SingleQueryBuilderType: SDQueryBuilder<M>;
+
+    delete(): this['SingleQueryBuilderType'] & M['QueryBuilderType']['SingleQueryBuilderType'];
+    hardDelete(): this['SingleQueryBuilderType'] & M['QueryBuilderType']['SingleQueryBuilderType'];
+    undelete(): this['SingleQueryBuilderType'] & M['QueryBuilderType']['SingleQueryBuilderType'];
+    whereDeleted(): this['SingleQueryBuilderType'] & M['QueryBuilderType']['SingleQueryBuilderType'];
+    whereNotDeleted(): this['SingleQueryBuilderType'] & M['QueryBuilderType']['SingleQueryBuilderType'];
+  }
```

Comment:
I manually used your PR in my project, but I made the types for the SDQueryBuilder class as such:
```
    import { Model, QueryBuilder } from 'objection';

    class SDQueryBuilder<M extends Model> extends QueryBuilder<M, R> {
        ArrayQueryBuilderType!: SDQueryBuilder<M>;

        SingleQueryBuilderType!: SDQueryBuilder<M>;

        NumberQueryBuilderType!: SDQueryBuilder<M, number>;

        PageQueryBuilderType!: SDQueryBuilder<M, Objection.Page<M>>;

        delete(): this['NumberQueryBuilderType'];
        hardDelete(): this['NumberQueryBuilderType'];
        undelete(): this['NumberQueryBuilderType'];
        whereDeleted(): this['ArrayQueryBuilderType'];
        whereNotDeleted(): this['ArrayQueryBuilderType'];
    }
```

Reply:
Looks cleaner. If it works then all's good :+1:

Reply:
Just updated the comment, we didn't need the `SingleQueryBuilderType` and the `PageQueryBuilderType` so I removed them. We would need them if the plugin added methods like `.findById()` and `.page()`. 

`.delete()`, `.hardDelete()` and `undelete()` internally use `.patch()`, which returns the `NumberQueryBuilder`.

`.whereDeleted()` and `.whereNotDeleted()` internally use `.where()` and `.whereNot()`, which returns the `ArrayQueryBuilder`